### PR TITLE
Wae demo paths

### PR DESCRIPTION
**Pull request by guilhermedemouraa**

### CODE REVIEW

**Comment by guilhermedemouraa** [comment]
I think that it would be cool to have this example on our website to teach folks how to create their own custom paths. I like the idea of sharing our own path used at the WAE. But I'm open to suggestions. One thing I thought that would be cool is we take in as args row_length, row_spacing, and n_rows. So we could just make something more general...

File: py/farm_ng/track/utils.py [hunk 1]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
...
+            counter += 1
+            remaining_distance -= segment_distance
+
+        # Rename the last pose to the desired name
+        segment_poses[-1].frame_b = next_frame_b
+        self.track_waypoints.extend(segment_poses)
+        self._segment_indices.append(len(self.track_waypoints))
+        self._loaded = False
+
+    def create_ab_segment(self, next_frame_b: str, final_pose: Pose3F64, spacing: float = 0.1) -> None:
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:161]: `create_straight_segment` and `create_ab_segment` seem to do very similar things. Suggest to have one call the other, or have a private method that both use

File: py/farm_ng/track/utils.py [hunk 2]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
...
+        """Initialize the TrackBuilder."""
+        self.clients = clients if clients is not None else {}
+        self._start: Pose3F64()
+        self.track_waypoints: list[Pose3F64] = []
+        self._segment_indices: list[int] = [0]
+        self._loaded: bool = False
+
+        if start is not None:
+            self._start = start
+            self.track_waypoints = [start]
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:72]: Can we just require a start on creation?

Other methods assume there are already waypoints (e.g., `segment_poses: list[Pose3F64] = [self.track_waypoints[-1]]`, and being able to cause a crash by calling these methods after the default constructor seems unsafe.

```suggestion
    def __init__(self, start: Pose3F64, clients: dict[str, EventClient] | None = None) -> None:
        """Initialize the TrackBuilder."""
        self.clients = clients if clients is not None else {}
        self._start: Pose3F64() = start
        self.track_waypoints: list[Pose3F64] = [start]
        self._segment_indices: list[int] = [0]
        self._loaded: bool = False
```

File: py/farm_ng/track/utils.py [hunk 3]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
     filter_track: FilterTrack = proto_from_json_file(track_path, FilterTrack())
     track: Track = filter_track_to_track(filter_track)
     proto_to_json_file(track_path, track)
+
+
+class TrackBuilder:
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:59]: I'd prefer to move this class to a file in `py/examples/track_planner/track_builder.py` until it is more mature and used in more places than this one example.

File: py/farm_ng/track/utils.py [hunk 4]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
...
+    """A class for building tracks."""
+
+    def __init__(self, clients: dict[str, EventClient] | None = None, start: Pose3F64 | None = None) -> None:
+        """Initialize the TrackBuilder."""
+        self.clients = clients if clients is not None else {}
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:64]: Are there any clients besides the filter client going to be used in this class? If the only client is the filter client, and it is only used in the one `get_pose` function, then this should be simplified by removing self.clients as a class member

File: py/farm_ng/track/utils.py [hunk 5]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
...
+            path (Path): The path of the file to import.
+        """
+        loaded_track = proto_from_json_file(path, Track())
+        self.track = loaded_track
+
+    def merge_tracks(self, track_to_merge: Track, threshold: float = 0.5) -> bool:
+        """Merge a track with the current track.
+
+        Args:
+            track (Track): The track to merge.
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:340]:
```suggestion
            track_to_merge (Track): The track to merge.
```

File: py/farm_ng/track/utils.py [hunk 6]
```diff
@@ -45,3 +54,314 @@ def update_filter_track(track_path: Path) -> None:
...
+        """
+        # TODO: Check if we can invert the track_to_merge and append it to the current track
```
Comment (Hackerman342) [inline py/farm_ng/track/utils.py:342]: Not entirely sure what this means, but consider addressing the TODO before merge

```suggestion
        # TODO: Check if we can invert the track_to_merge and append it to the current track
```

File: py/farm_ng/track/utils.py (hunk 1 above)
Comment (Hackerman342) [reply py/farm_ng/track/utils.py:161]: Actually there's a lot of boilerplate with all of `create_straight_segment`, `create_ab_segment`, `create_turn_segment`, & `create_arc_segment` that can be refactored to make these methods more concise and readable

**Review by Hackerman342 — COMMENTED** [review]
If you want to move the `TrackBuilder` class into `py/examples/track_planner/track_builder.py` (as noted in one comment), then I'll approve right away. If you want this in the actual farm-ng-amiga module, then we need to harden it and consider how it will be used in the codebase / apps / etc.. This is because we will want to avoid breaking backwards compatibility once it is merged, and especially once it is released

File: py/farm_ng/track/utils.py (hunk 3 above)
Comment (guilhermedemouraa) [reply py/farm_ng/track/utils.py]: Done here: 4eb3337009bdca9774f93ef55ea7a5be03c28f35

File: py/farm_ng/track/utils.py (hunk 4 above)
Comment (guilhermedemouraa) [reply py/farm_ng/track/utils.py]: No, it's only the filter. I removed the check the function get_pose() from the class. I guess it makes more sense to have the class only do track planning.

File: py/farm_ng/track/utils.py (hunk 2 above)
Comment (guilhermedemouraa) [reply py/farm_ng/track/utils.py]: Sure, we can require a start on creation. I created a function in the `main.py` example to query the current robot pose, and modified the class in this commit: 3ca6744c813c1cb28c62234e2c99f0ca8846c445

File: py/farm_ng/track/utils.py (hunk 1 above)
Comment (guilhermedemouraa) [reply py/farm_ng/track/utils.py]: Boilerplate code optimized here: 4eb3337009bdca9774f93ef55ea7a5be03c28f35

**Review by Hackerman342 — APPROVED** [review]

**Pull request merged**